lib/mpl_toolkits/mplot3d/art3d.py

```diff
@@ -649,37 +661,39 @@ def do_3d_projection(self, renderer):
             self.update_scalarmappable()
             self._facecolors3d = self._facecolors
 
-        txs, tys, tzs = proj3d._proj_transform_vec(self._vec, renderer.M)
-        xyzlist = [(txs[sl], tys[sl], tzs[sl]) for sl in self._segslices]
+        psegments = proj3d._proj_transform_vectors(self._segments, renderer.M)
+        num_faces = len(psegments)
 
         # This extra fuss is to re-order face / edge colors
         cface = self._facecolors3d
         cedge = self._edgecolors3d
-        if len(cface) != len(xyzlist):
-            cface = cface.repeat(len(xyzlist), axis=0)
-        if len(cedge) != len(xyzlist):
+        if len(cface) != num_faces:
+            cface = cface.repeat(num_faces, axis=0)
+        if len(cedge) != num_faces:
             if len(cedge) == 0:
                 cedge = cface
             else:
-                cedge = cedge.repeat(len(xyzlist), axis=0)
+                cedge = cedge.repeat(num_faces, axis=0)
 
-        # sort by depth (furthest drawn first)
-        z_segments_2d = sorted(
-            ((self._zsortfunc(zs), np.column_stack([xs, ys]), fc, ec, idx)
-             for idx, ((xs, ys, zs), fc, ec)
-             in enumerate(zip(xyzlist, cface, cedge))),
-            key=lambda x: x[0], reverse=True)
+        face_z = self._zsortfunc(psegments[..., 2], axis=-1)
+        if isinstance(face_z, np.ma.MaskedArray):
+            # NOTE: Unpacking .data is safe here, because every face has to
+            #       contain a valid vertex.
+            face_z = face_z.data
+        face_order = np.argsort(face_z, axis=-1)[::-1]
 
-        segments_2d = [s for z, s, fc, ec, idx in z_segments_2d]
         if self._codes3d is not None:
-            codes = [self._codes3d[idx] for z, s, fc, ec, idx in z_segments_2d]
+            segments_2d = [s.compressed().reshape(-1, 2)
```

Comment:
`compressed` is going to do something bad if someone only one component was masked (eg, it was `nan` at construction).

Safer would be:
```suggestion
            segments_2d = [s.data[~s.mask.any(axis=-1)]
```

Reply:
How would you like to interpret the data if one component was masked? That makes no sense because you get one point with only a single coordinate.

Reply:
Oh you mean that multiple similar errors like this could cancel out? That is true, but it's such a strong internal invariant violation that I'm not sure if what your solution does (i.e. just take masked data without asking too many questions) is better? Another alternative would be to do `s.data[~s.mask[..., 0]]`, as the last dimension of mask is always supposed to be constant.

Reply:
You're right it makes no sense, but the way you have it written the behavior would be to interpret `[[x1, y1], [x2, --], [x3, --], [x4, y4]]` as `[[x1, y1] ,[x2, x3], [x4, y4]]`, which is going to be very confusing to the user as it combines adjacent points.

Reply:
Ok, I guess using `mask.all` might be more reasonable. Would that be ok with you?

Reply:
I've looked into it, and I'm not sure if we should be doing anything like that. Note that right below this line there is another comprehension that builds up a list of codes, each with length that is supposed to be equal to the length of this segment. So we can't just drop vertices arbitrarily and the only way to make this sane is to assume that the invariant I've mentioned holds. What do you think?